add vpc nat gateway module.

jonhadfield opened the pull request:
This adds the functionality to manage a VPC Managed NAT Gateway.

Previous PR https://github.com/ansible/ansible-modules-extras/pull/1438 closed as github won't allow me to re-open after a forced push.

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,303 @@
+    required: false
+    default: 300
+  region:
+    description:
```
don't add region. extend with 'ec2' doc fragment instead

jonhadfield commented:
@wimnat I've corrected and squashed previous commits.

robynbergeron commented:
Thanks @jonhadfield for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

viper233 commented:
```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"allocation_id": "eipalloc-XXXXXXX", "aws_access_key": "ASIAJR62UYMGF4TYUJTA", "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "ec2_url": null, "nat_gateway_id": null, "profile": null, "region": null, "security_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "state": "present", "subnet_id": "subnet-XXXXXXX", "validate_certs": true, "wait": true, "wait_timeout": 300}, "module_name": "ec2_vpc_nat_gateway"}, "msg": "allocation: eipalloc-XXXXXXX does not exist."}
```

I'm not able associate with an eip allocation when assuming a role in another account. I don't have any problem creating the eip and it seems valid, correct id and VPC.

Any idea what could be wrong? I'm getting a similar issue with the #1446 nat gateway facts module not authenticating. Other AWS modules work fine, ec2_vpc_net_facts, ec2_vpc_subnet_facts.

Thanks!

viper233 commented:
Nevermind, this has been fixed in ansible/ansible/pull/14633

khogeland commented:
I think this is missing something, or maybe I'm missing something... currently, the method of identification (and therefore idempotency) is EIP allocation. EIPs are by design not identifiable (allocate when needed, or take an available allocated IP if possible), so they are unusable as ID. It seems there needs to be a field like `multiple_gateways_allowed` which would let you enforce a single NAT gateway per subnet or VPC.

jonhadfield commented:
Thanks for the feedback @khogeland.
To (idempotently) enforce the presence of a NAT gateway in a particular subnet with a specific EIP you must provide at a minimum the subnet-id and the allocation-id of the EIP.
I intentionally don't auto-allocate an EIP when creating a NAT gateway as I believe that's the job of the ec2_eip module. This not only forces idempotence (you have to specify an existing allocation-id) but avoids duplicating functionality.
I managed to get https://github.com/ansible/ansible-modules-core/pull/2753 merged so that the allocation-id would be returned for a successful VPC EIP allocation.

khogeland commented:
Right, it's the "with a specific EIP" part that is a problem. The `ec2_eip` module is _not_ idempotent, it only returns an unused EIP, allocating a new one if all current EIPs are associated. What you're suggesting requires users to do backflips to not create another EIP (which would cause another NGW to be created). That's backwards - you shouldn't be managing EIPs, the EIPs should only be created if needed. The user flow should like this this:
`user requests NGW -> no NGWs in subnet -> module creates NGW and associates new IP`
`user requests NGW -> already an NGW in subnet -> module returns existing NGW`

jonhadfield commented:
It is possible to have more than one NAT gateway per subnet, so to ensure idempotency of ensuring one exists, you must provide a combination of allocation-id and subnet-id. So the second flow would prevent more than one being created.
I imagine the flows should be:
    `user requests NGW with alloc-id and subnet id -> doesn't exist -> module creates`
    `user requests NGW with alloc-id and subnet id -> does exist -> shows un-changed`
I guess there could be a third flow:
    `user requests NGW with subnet-id -> module allocates an EIP and creates another NAT GW`
but this wouldn't be idempotent as it would just keep adding NAT GWs until you hit your EIP limit

khogeland commented:
I don't agree, but I don't have a better solution. Maybe an `exact_count` parameter similar to the `ec2` module? Unsure why Amazon decided that NGWs shouldn't be tagged. :confused:

jonhadfield commented:
I guess I should have put _'I imagine the flows can only be'_ as it's not really a preference but a limitation.
I keep raising the tagging issues: limited to 10, lack of them on sec group rules, etc. with our AWS contact, but I only get shrugs as a response.

gaieges commented:
Where does this stand?  Is it waiting on further review or [tag] modifications from Amazon?

jonhadfield commented:
@gaieges The module has been created and reviewed by myself and @Etherdaemon and tested by ourselves and other community members: https://github.com/ansible/ansible-modules-extras/pull/1438
The discussion above was a concern/preference over idempotency but I don't think there is a clear case for further changes.
We're now just waiting for it to be merged.

aioue commented:
Related https://github.com/ansible/ansible-modules-extras/pull/1882

aioue commented:
@jonhadfield @Etherdaemon is it possibly to add this file to my project and reference it someone in order to test/use it before it is merged into 2.1?

Etherdaemon commented:
@aioue for sure - I already do this now for modules that I have written and hasn't been merged yet due to reviews etc. Just add it to your library folder and ensure it includes the contributor.

EG:

```
├── README.md
├── STRUCTURE_GUIDELINE.md
├── ansible.cfg
├── inventory
├── library
  ├── ec2_vpc_nat_gateway.py
├── roles
│   ├── allinone
│   │   ├── defaults
│   │   │   └── main.yml
│   │   └── tasks
│   │       ├── main.yml
└── playbook_a.yml
```

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,298 @@
+    type: string
+    sample: "nat-0d1e3a878585988f8"
+'''
+
```
Need some RETURN documentation

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,298 @@
+    HAS_BOTO_3 = False
+from distutils.version import LooseVersion
+if LooseVersion(botocore.__version__) < LooseVersion("1.3.14"):
+    HAS_SUFFICIENT_BOTOCORE = False
```
As per guidelines...

> Use feature testing (e.g. hasattr('boto.module', 'shiny_new_method')) to check whether boto supports a feature rather than version checking

jonhadfield replied:
I think I'm forced to do version checking in this scenario. If you don't have the version of botocore that supports NAT GWs, you get the following: 
"'EC2' object has no attribute 'describe_nat_gateways'"
A quick look through boto3 and botocore code suggests the only potential check is to look at the json files in botocore.

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,298 @@
+        wait = module.params.get('wait')
+        start_time = datetime.datetime.utcnow()
+        if wait:
+            wait_timeout = time.time() + module.params.get('wait_timeout')
```
You should use http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Waiter.NatGatewayAvailable rather than your own custom wait

jonhadfield replied:
I've had a play and I don't think the waiter is the best solution for the following reasons (aside from poor documentation):
1. The failure message is too basic to be useful, e.g.: _"Waiter NatGatewayAvailable failed: Waiter encountered a terminal failure state"_ At least using a custom wait allows for specific failure messages to be returned to the user.
2. There is no waiter for 'NatGatewayDeleted', so I'd still be forced to implement a custom wait for deletion. I don't think it's safe 'not' to wait in all circumstances as it may be followed by an action that requires it to be absent, e.g. re-use of the EIP.

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,298 @@
+                    if present_status.get('state') == 'failed' and create_time.replace(tzinfo=None) >= start_time:
+                        module.fail_json(msg="Failed to create VPC NAT Gateway")
+                    elif present_status.get('state') == 'available':
+                        nat_gateway_id = result['NatGateway']['NatGatewayId']
```
Have a look at using the helper function from module utils "camel_dict_to_snake_dict" and return all the parameters that boto3 returns not just gateway id - i guarantee someone will want one of them later!

jonhadfield replied:
Done.

```
ok: [localhost] => {
    "nat_gateway_pre": {
        "changed": true,
        "nat_gateway": {
            "create_time": "2016-05-02T21:21:50.683000+00:00",
            "nat_gateway_addresses": [
                {
                    "allocation_id": "eipalloc-784f391d"
                }
            ],
            "nat_gateway_id": "nat-03afc583f2b2962b6",
            "state": "pending",
            "subnet_id": "subnet-16640561",
            "vpc_id": "vpc-6a31bc0f"
        }
    }
}
```

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,298 @@
+                    time.sleep(5)
+            module.fail_json(msg="Waited too long for VPC NAT Gateway to be created.")
+        else:
+            module.exit_json(changed=True)
```
This statement should return the same values as your other exit_json statement above.

Maybe rework the flow slightly so there is only one exit_json statement

jonhadfield replied:
Reworked to use single exit_json statement. Unfortunately, nat_gateway_id will only be available if user decided to wait for result.

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,298 @@
+        module.exit_json(**results)
+    except botocore.exceptions.ClientError as ce:
+        if "NatGatewayMalformed" in ce.message:
+            module.fail_json(msg="Invalid NAT Gateway ID")
```
Don't look for a string in message.  Use error code provided instead.

E.g...
if e.response['Error']['Code'] == 'bla':

jonhadfield replied:
Will do.

wimnat commented:
needs_revision

jonhadfield commented:
Will sort out today. Thanks.

aioue commented:
Tested this module locally. Worked but required tweak. `import boto` above the `import boto3` (line 101).

Without it:

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "<stdin>", line 2505, in <module>
  File "<stdin>", line 280, in main
  File "<stdin>", line 2414, in get_aws_connection_info
NameError: global name 'boto' is not defined

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "ec2_vpc_nat_gateway"}, "module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 2505, in <module>\n  File \"<stdin>\", line 280, in main\n  File \"<stdin>\", line 2414, in get_aws_co
nnection_info\nNameError: global name 'boto' is not defined\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
```

aioue commented:
It would be great to be able to pass the actual EIP rather than the `allocation id`.

jonhadfield commented:
Thanks @aioue 
RE: boto import:
- This is an issue with the 'module_utils/ec2.py' module, where it makes use of boto(version 2) module which this Ansible module doesn't require. I've not checked the latest code but, IMHO, if boto3 is available (which I check for), it should instead be used to establish the connection. 

RE: EIP/Allocation ID:
- This is on purpose as boto3 only accepts the allocation id as the identifier for the EIP (http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Client.create_nat_gateway) and I suspect it's for good reason as, elsewhere, public IPs are valid. Rather than accept the public IP (and have to perform a lookup) I got a change merged to the ec2_eip module that makes the registered output include the allocation id. 

More inclined to keep it simple than accommodate all use cases, but not too difficult to address both issues.

aioue commented:
@jonhadfield ok re:allocation ID, working with the API makes the most sense.

re: boto: I have boto3 installed and actually use a modified ec2.py which imports it for the inventory. However the error above occurred and went away following and explicit import of boto.

jonhadfield commented:
@aioue a comparison between the latest ec2.py in devel and an early v2 version I normally use shows differences in the imports. If you could let me know the tag and the custom ec2.py I can investigate. Suspect there's some import weirdness going on.
An explicit import of boto(2) feels like a workaround for another issue, but not too worried if it's only short term and fixes it for everyone.

jonhadfield commented:
@wimnat I've added comments and pushed changes based on your suggestions.
@Etherdaemon I've added you as a collaborator on my fork in case I'm not available and you want to progress.

gregdek commented:
Thanks @jonhadfield, looking forward to some re-reviews here.

aioue commented:
Seems to have regressed from 7 days ago.

```
failed: [localhost] (item={u'tags': {u'role': u'bastion,natgateway,elb', u'type': u'public', u'Name': u'prod2_prod2_a_public_bastion', u'env': u'prod2'}, u'availability_zone': u'eu-west-1a', u'default_for_az': u'false', u'state': u'available', u'vpc_id': u'vpc-sdfsdfsdff', u'cidr_block': u'172.17.11.0/24', u'available_ip_address_count': 251, u'id': u'subnet-878a55e3', u'map_public_ip_on_launch': u'false'}) => {"failed": true, "invocation": {"module_args": {"allocation_id": "eipalloc-345345", "aws_access_key": null, "aws_secret_key": null, "ec2_url": null, "nat_gateway_id": null, "profile": null, "region": null, "security_token": null, "state": "present", "subnet_id": "subnet-878a55e3", "validate_certs": true, "wait": true, "wait_timeout": 300}, "module_name": "ec2_vpc_nat_gateway"}, "item": {"availability_zone": "eu-west-1a", "available_ip_address_count": 251, "cidr_block": "172.17.11.0/24", "default_for_az": "false", "id": "subnet-4534534", "map_public_ip_on_launch": "false", "state": "available", "tags": {"Name": "prod2_prod2_a_public_bastion", "env": "prod2", "role": "bastion,natgateway,elb", "type": "public"}, "vpc_id": "vpc-345345345"}, "msg": "global name 'camel_dict_to_snake_dict' is not defined"}
```

jonhadfield commented:
Thanks for testing @aioue.
The issue is that it's been suggested I use a method called 'camel_dict_to_snake_dict' from the common module_utils/ec2 module but that probably hasn't made it into the ansible version you are testing against.
Please could you retry with: v2.1.0.0-0.1.rc1?

aioue commented:
@jonhadfield I'm rather tight for time unfortunately. Happy to re-test when 2.1 is available on homebrew.

Alternatively, is there a library path that I can put the [updated ec2.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/ec2.py) in to test?

aioue commented:
@jonhadfield dropping the linked `ec2.py` into `./library` fixed the error. Caused the existing deployment (which actually appeared to have worked despite the errors) to be changed, but when run again was idempotent.

It still requires the `import boto` to work though.

jonhadfield commented:
@aioue If it gets merged, I imagine it will be against Ansible 2.2 which will hopefully have that issue fixed (making use of boto in ec2.py despite not being explicitly imported). If not, it can simply be added, whenever, at the risk of duplication.
Thanks again.

wimnat commented:
@jonhadfield can you squash the commits?

jonhadfield commented:
@wimnat Done.

ryansb commented:
After reviewing & testing this, I think the idempotency problem in this (naive) playbook will end with a lot of users shooting themselves in the foot:

```
---
- hosts: localhost
  tasks:
  - name: allocate EIP
    ec2_eip:
      in_vpc: true
      region: us-west-2
      state: present
    register: eip
  - name: Test NAT gateway
    ec2_vpc_nat_gateway:
      state: present
      subnet_id: subnet-38179c5c
      region: us-west-2
      allocation_id: "{{ eip.allocation_id }}"
```

The above playbook will provision a new EIP and NAT Gateway every time, and the only way to make it _not_ do that is to hardcode the EIP allocation ID for the gateway, and have the IP allocated by hand or in a separate playbook. 

Because the EIP module is not idempotent (for unfortunately required reasons due to the semantics of detached/attached state) the options for this module are limited. I think the only way to make this work the way most users would expect would be like: 

```
  - ec2_vpc_nat_gateway:
      state: present
      subnet_id: subnet-38179c5c # required
      allocation_id: "{{ eip.allocation_id }}" # optional, would auto-allocate or use detached EIP
      exact_count: 1 # optional, defaulting to 1
```

That way a user that created a NAT Gateway wouldn't be surprised after 5 playbook runs to find that they have 5 gateways (only one of which is actually used).

jonhadfield commented:
Thanks for the review @ryansb.
I'll have a play with that scenario and suggested workaround/fix and come back to you.

jonhadfield commented:
@ryansb Wherever exact_count is mentioned in existing roles it is used, not just to provide idempotency, but to terminate instances if too many have previously been created. 
If a NAT Gateway were terminated (to reduce to a specified exact count) then it could lead to a route table being invalid and have a greater knock on effect than ending up with duplicates.
So, would it be preferable to:
- use exact_count to specify a threshold after which no more will be created, i.e. do nothing if count >= exact_count?
- use exact_count to force a specific count (terminate oldest?)
- use a different var name to force a threshold that doesn't conflict with existing usage of exact count?

ryansb commented:
I see how that could be a problem, since ansible wouldn't know _which_ to terminate since one or more of them may be in the route tables for different VPC's. I think terminating the oldest would be reasonable behavior. As long as it's not random, and is documented as to what happens when there are too many, then that would be ok. 

What do you mean by use a different var name to force a threshold? You mean for the parameter?

jonhadfield commented:
Yes, the parameter name. In the case where exact_count didn't match the behaviour of other modules, then it might make sense to rename it.
I'll have a stab at adding exact_count as a parameter with (documented) default behaviour as removing oldest.

jonhadfield commented:
@ryansb I decided against 'exact_count' as it isn't logical in certain scenarios. For example, if user puts exact_count: 2 and none existed, then they have only provided one allocation id, so the state cannot be attained.

Instead, I've added a parameter called 'max_count' with a default value of 1 which handles the following scenarios:
- max_count = 1 and current count = 0, module creates NAT GW
- max_count = 1 and current count = 1, if the subnet and allocation ids are the same, then unchanged. If subnet id is the same and allocation id is different, then fail with relevant message.
- max_count = 1 and current count = 2, then fail with relevant message

I've re-tested with your playbook (above) and with a number of other tasks in the following playbooks:
Successes: https://gist.github.com/jonhadfield/64967ac919e2e020c020bc8cb7c23474
Failures: https://gist.github.com/jonhadfield/603bd6b09abdb352bfc2eba359da0a77

Btw, Not sure why travis is failing, but seems to be affecting other PRs also.

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+                   allocation_id=nat_gateway_address.get('AllocationId'))
+    ec2_client.delete_nat_gateway(NatGatewayId=nat_gateway_id)
+    wait = module.params.get('wait')
+    if wait:
```
Can i suggest using a boto3 waiter here rather than 'get_nat_gateway_status_list'?

http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Waiter.NatGatewayAvailable

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+        start_time = datetime.datetime.utcnow()
+        results = dict(changed=True)
+        if wait:
+            wait_timeout = time.time() + module.params.get('wait_timeout')
```
Same as above regarding waiter http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Waiter.NatGatewayAvailable

jonhadfield commented:
As commented above...

> 'I've had a play and I don't think the waiter is the best solution for the following reasons (aside from poor documentation):
> 1. The failure message is too basic to be useful, e.g.: "Waiter NatGatewayAvailable failed: Waiter encountered a terminal failure state" At least using a custom wait allows for specific failure messages to be returned to the user.
> 2. There is no waiter for 'NatGatewayDeleted', so I'd still be forced to implement a custom wait for deletion. I don't think it's safe 'not' to wait in all circumstances as it may be followed by an action that requires it to be absent, e.g. re-use of the EIP.'

I've re-checked and they still haven't included more waiters: http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#waiters

aioue commented:
The latest commit is still working for me as long as I import boto. 👍

jonhadfield commented:
Thanks @aioue. 
Please could you clarify which version of Ansible you're using? I believe the issue in module_utils/ec2.py has been updated, in devel at least, to remove hard dependency on boto(2) if boto3 is installed.

aioue commented:
@jonhadfield 2.0.2.0 with the latest devel ec2.py in my library folder.

jonhadfield commented:
@aioue I've re-tested and it will only work against Ansible 2.1 and newer as previously:
1) module_utils/ec2.py wouldn't check for boto3 before making use of it
2) I'm making use of the 'camel_dict_to_snake_dict' function in module_utils/ec2.py that wasn't introduced until 2.1
That would explain why you need to backport that functionality for this module to work.

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+        if ce.response['Error']['Code'] == "NatGatewayNotFound":
+            return [{'state': 'absent'}]
+        else:
+            module.fail_json(msg=ce.message)
```
Better exception info as per guidelines:

`module.fail_json(msg=ce.message, **camel_dict_to_snake_dict(ce.response))`

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+                module.fail_json(msg="Waited too long for VPC NAT Gateway to be created.")
+        module.exit_json(**results)
+    except botocore.exceptions.ClientError as ce:
+        module.fail_json(msg=ce.message)
```
Better exception info as per guidelines:

`module.fail_json(msg=ce.message, **camel_dict_to_snake_dict(ce.response))`

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+        if ce.response['Error']['Code'] == "NatGatewayMalformed":
+            module.fail_json(msg="Invalid NAT Gateway ID")
+        else:
+            module.fail_json(msg=ce.message)
```
Better exception info as per guidelines:

`module.fail_json(msg=ce.message, **camel_dict_to_snake_dict(ce.response))`

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+    try:
+        ec2_client.describe_addresses(AllocationIds=[module.params.get('allocation_id')])
+    except botocore.exceptions.ClientError:
+        module.fail_json(msg="allocation: %s does not exist." % allocation_id)
```
Probably not wise to assume this ClientError is because of allocation not existing.  Can you check the response code?

Also as per the others...

Better exception info as per guidelines:

`module.fail_json(msg=ce.message, **camel_dict_to_snake_dict(ce.response))`

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+    try:
+        ec2_client.describe_subnets(SubnetIds=[subnet_id])
+    except botocore.exceptions.ClientError:
+        module.fail_json(msg="subnet: %s does not exist." % subnet_id)
```
Same as other comment above

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+                module.exit_json(dict(change=True))
+    except botocore.exceptions.ClientError as ce:
+        if ce.response['Error']['Code'] == "NatGatewayLimitExceeded":
+            module.fail_json(msg="The NAT Gateway limit has been exceeded.")
```
Prefer to go with boto3 message.  Not sure what it says though! Is it not good enough?

Better exception info as per guidelines:

`module.fail_json(msg=ce.message, **camel_dict_to_snake_dict(ce.response))`

wimnat commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,356 @@
+    except botocore.exceptions.NoCredentialsError:
+        module.fail_json(msg="Unable to locate AWS credentials.")
+    except Exception as ce:
+        module.fail_json(msg=ce.message)
```
Don't catch any Exception.  Particularly because a non-boto exception probably wont have a message attribute

jonhadfield commented:
Hi @wimnat, I've made the requested changes and re-tested success and failure scenarios.
Travis build failing due to unrelated change in ec2_vpc_vgw.py.

gregdek commented:
Thanks @jonhadfield for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

jonhadfield commented:
@gregdek It failed to build due to an issue with an unrelated module that's now fixed.
I've rebased which has forced a rebuild and it's now green.

gregdek commented:
Thanks @jonhadfield -- it appears to be related to "old" PRs, for some value of "old" that we can't quite nail down. I hesitate to say "just rebase and it'll probably go away," but that seems to be the answer in most cases.

wimnat commented:
shipit works_for_me

ryansb commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,347 @@
+        # Check max_count has not been exceeded before attempting to create
+        if module.params.get('max_count') < num_nat_gateways:
+            module.fail_json(
+                msg="The maximum desired number of NAT Gateways specified in "
```
This shouldn't fail when the right number exist - that's a desired scenario.

ryansb commented on cloud/amazon/ec2_vpc_nat_gateway.py:
```diff
@@ -0,0 +1,347 @@
+                                                                 {'Name': 'state',
+                                                                  'Values': ['available', 'pending',
+                                                                             'deleting']}])
+        num_nat_gateways = len(nat_gateways.get('NatGateways')) if nat_gateways.get('NatGateways') else 0
```
Can you use `len(nat_gateways.get('NatGateways', []))` instead to avoid the inline `if`?

ryansb commented:
I'm not quite happy with the behavior yet. If the right number of instances _do_ exist, the module fails. My play:

```
  - name: allocate EIP
    ec2_eip:
      in_vpc: true
      region: us-west-2
      state: present
    register: eip
  - name: Test NAT gateway
    ec2_vpc_nat_gateway:
      state: present
      subnet_id: "{{ my-subnet }}"
      region: us-west-2
      allocation_id: "{{ eip.allocation_id }}"
      max_count: 1
```

The first time it works great, but all subsequent runs fail with:

```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"allocation_id": "eipalloc-1cf9d478", "aws_access
_key": null, "aws_secret_key": null, "ec2_url": null, "max_count": 1, "nat_gateway_id": null, "profile": null, "region": "us-west-2", "security_to
ken": null, "state": "present", "subnet_id": "subnet-9f18bdc6", "validate_certs": true, "wait": true, "wait_timeout": 300}, "module_name": "ec2_vp
c_nat_gateway"}, "msg": "The maximum desired number of NAT Gateways specified in max_count (default: 1) has already been met."}
```

jonhadfield commented:
The logic flow is:
- Check number of NAT GWs in specified subnet (if max_count exceeded - fail)
- If number not exceeded, check to see if specified NAT GW already exists (if so, exit with  changed=False)
- If the NAT gateway doesn't already exist, then check that creation of a new one won't exceed max count
  ...

On your second run, you've specified a different EIP, so it will attempt to create a new NAT GW but fail because it's reached max_count. If the EIP was the same, then it would exit unchanged.

jonhadfield commented:
Removed 'inline if' as suggested by @ryansb, re-tested scenarios and re-based.

```
-        num_nat_gateways = len(nat_gateways.get('NatGateways')) if nat_gateways.get('NatGateways') else 0
+        num_nat_gateways = len(nat_gateways.get('NatGateways', []))
```

ryansb commented:
I disagree with the logic in the module. Since you have to specify an allocation ID, how does having a `max_count` of greater than one make sense? 

To have a playbook run on the first try, I have to:
1. Provision an EIP as part of the playbook
2. Store the EIP ID somewhere manually
3. Provision a NAT Gateway with that ID

I think a sensible tweak would be:
1. Check number of NAT GWs in specified subnet (if max_count equal to current number - exit changed=False)
2. If max_count exceeded - exit with fail
3. Check NAT GW (by EIP allocation ID) if so, exit with changed=False
4. If the NAT gateway doesn't already exist, and the new one won't exceed max_count, then  then check that creation of a new one won't exceed max count, create and exit changed=True

jonhadfield commented:
max_count is to limit the number of NAT Gateways per subnet, not per allocation ID (specified in document string).
My concern with item 1 you put above is that a user is defining a state based on subnet and allocation id combined. So, in a scenario where they said max_count: 1, subnet_id: s1, allocation_id: a1, and the current state was having a single NAT Gateway in the subnet s1 with a allocation_id a2, the desired state has not been met. 
I could be more specific by changing the parameter to max_count_per_subnet, or modify as you suggest and keep an eye on issues/feedback if it gets merged? Just worried this might be adding further ambiguity (of what constitutes desired state) for a separate EIP allocation issue.

gregdek commented:
@jonhadfield A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

jonhadfield commented:
No change made yet, but looking for feedback to previous comment.
ready_for_review

gregdek commented:
Thanks @jonhadfield for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

badmojo76 commented:
Thanks @jonhadfield for putting this module together - and apologies in advance for the long post.  

Just a thought - it appears that the "clientToken" parameter is intended to combat the issues with idempotency...  So why not introduce it as a variable and let AWS's API handle any Idempotency mismatch issues (i.e., overlapping clientToken names in the same region)?

Based on AWS docs, the clientToken should only allow the request to be executed the FIRST time, given that no other settings change (i.e., subnet, EIP allocation, etc.), with each subsequent run simply returning the same values, but not actually taking action.  Looks accurate, based on my testing:

AWS CLI command:

```
$ aws ec2 create-nat-gateway --subnet-id subnet-5defc605 --allocation-id eipalloc-3fadd858 --client-token this-is-my-nat-version-1
```

First run:

```
{
    "ClientToken": "this-is-my-nat-version-1",
    "NatGateway": {
        "NatGatewayAddresses": [
            {
                "NetworkInterfaceId": "eni-b36459ab",
                "AllocationId": "eipalloc-3fadd858",
                "PrivateIp": "10.200.5.96"
            }
        ],
        "VpcId": "vpc-f26de295",
        "State": "pending",               # Note the state of pending
        "NatGatewayId": "nat-00ee0eb8dc528ed3a",
        "SubnetId": "subnet-5defc605",
        "CreateTime": "2016-07-13T00:13:00.680Z"
    }
}
```

and once it is out of "pending" state, another run with the exact same configuration results in:

```
{
    "ClientToken": "this-is-my-nat-version-1",
    "NatGateway": {
        "NatGatewayAddresses": [
            {
                "PublicIp": "52.201.56.124",           # Public IP shows up
                "NetworkInterfaceId": "eni-b36459ab",
                "AllocationId": "eipalloc-3fadd858",
                "PrivateIp": "10.200.5.96"         
            }
        ],
        "VpcId": "vpc-f26de295",
        "State": "available",                          # NGW is available
        "NatGatewayId": "nat-00ee0eb8dc528ed3a",
        "SubnetId": "subnet-5defc605",
        "CreateTime": "2016-07-13T00:13:00.680Z"
    }
}
```

If an attempt to modify the EIP, Subnet, etc. is made, and the clientToken is not modified, AWS will return (in this example, subnet), or if someone mistakenly names 2 NGWs in the same region the same, they'll get:

```
A client error (IdempotentParameterMismatch) occurred when calling the CreateNatGateway operation: nat-00ee0eb8dc528ed3a already created with the same token but different vpc/subnet
```

If the gateway is subsequently deleted via DeleteNatGateway, the clientToken is reset on the AWS side as soon as the NGW state goes to "Deleted".  So, for example, if a NGW is accidentally deleted, ansible will fail while the state is in "Deleting":

```
A client error (IdempotentParameterMismatch) occurred when calling the CreateNatGateway operation: nat-0c14bb00a3f077e92 was created with same token but is in an invalid state
```

...but will succeed and build a new NGW as soon as the old one goes to "Deleted" (a few mins).  

Taking @ryansb  example (modified for use with clientToken):

```
 - name: Test NAT gateway
    ec2_vpc_nat_gateway:
      state: present
      subnet_id: "{{ items.subnet_id }}"
      region: us-west-2
      allocation_id: "{{ items.eip_allocation_id }}"
      client_token: my-nat-gateway-{{ items.gateway_id }}
    with_items:
      - { gateway_id: 'dev', eip_allocation_id: '{{ eip1 }}', subnet_id: '{{ subnet1 }}' }
      - { gateway_id: 'test', eip_allocation_id: '{{ eip2 }}', subnet_id: '{{ subnet2 }}' }
      - { gateway_id: 'prod', eip_allocation_id: '{{ eip3 }}', subnet_id: '{{ subnet3 }}' }
```

and you'll get exactly 3 NGWs with 3 EIPs in 3 specific subnets, every time.  

Delete functionality would need to remain as-is with the nat-id, I think.  DescribeNatGateways doesn't return a clientToken, and you wouldn't want to create a NGW if it didn't exist, only to turn around and delete it, of course. 

No need for arbitrary counts (max or exact).  I would think the clientToken _should_ be required, but I may not be thinking of a case where that's not ideal... 

Thoughts?

K

gregdek commented:
@jonhadfield so it looks like we've got a situation where we've got two modules that do more or less the same things, and they're both close to the finish line: this one and #1882. An embarrassment of riches, in this case. Thanks to both you and @linuxdynasty, and apologies to both of you for not asking you to join forces sooner; we're still not very good at identifying potentially duplicated modules.

The one thing in favor of #1882, from my perspective, is that it appears to have tests with it. 

@jonhadfield could I ask you to take a look at this and #1882 and let me know if there are substantial differences/improvements with this PR that we could incorporate into #1882? Or, if you think this PR is superior to #1882, I'd like to hear that as well.

jonhadfield commented:
@gregdek I first submitted this module 30th December 2015 and then discovered @Etherdaemon had previously submitted one to perform the same task. We worked together and @Etherdaemon then generously closed their PR so we could expedite the inclusion of this functionality into Ansible.

I've had a brief look at https://github.com/ansible/ansible-modules-extras/pull/1882 and I think the main difference is that our submission is less complex. For example, I intentionally don't use a public IP as a parameter as I think it unnecessary, but you could also argue that some users may prefer to specify a public IP. I think both modules achieve the same goal and are well written, so it's more of a subjective choice for yourselves and the community.

I appreciate the inclusion of tests for https://github.com/ansible/ansible-modules-extras/pull/1882 is a definite advantage. I'd be happy to write them for this module also if it's decided to continue with my PR.

Also happy to close this and help out with @linuxdynasty's submission as I'm more interested in getting the functionality merged than attribution for my efforts (assuming @Etherdaemon feels the same way as https://github.com/ansible/ansible-modules-extras/pull/1882 is based off their original module).

gregdek commented:
After discussion, it's time to make a decision, and the decision I'm making is that we're going with #1882 because it has tests. @jonhadfield and @Etherdaemon, thanks to both of you for your help and your patience throughout this process.
